cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java

```diff
@@ -0,0 +1,272 @@
+import org.jasig.cas.support.pac4j.authentication.principal.ClientCredential;
+import org.jasig.cas.ticket.TicketGrantingTicket;
+import org.jasig.cas.web.support.WebUtils;
+
+import org.pac4j.core.client.BaseClient;
+import org.pac4j.core.client.Client;
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.client.Mechanism;
+import org.pac4j.core.context.J2EContext;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.core.exception.RequiresHttpAction;
+import org.pac4j.core.exception.TechnicalException;
+import org.pac4j.core.profile.CommonProfile;
+import org.pac4j.core.profile.ProfileHelper;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.webflow.action.AbstractAction;
+import org.springframework.webflow.context.ExternalContext;
+import org.springframework.webflow.context.ExternalContextHolder;
+import org.springframework.webflow.execution.Event;
+import org.springframework.webflow.execution.RequestContext;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import javax.validation.constraints.NotNull;
+
+/**
+ * This class represents an action to put at the beginning of the webflow.
+ * <p>
+ * Before any authentication, redirection urls are computed for the different clients defined as well as the theme,
+ * locale, method and service are saved into the web session.</p>
+ * After authentication, appropriate information are expected on this callback url to finish the authentication
+ * process with the provider.
+ *
+ * @author Jerome Leleu
+ * @author Longze Chen
+ * @since 4.1.0
+ */
+public final class ClientAction extends AbstractAction {
+
+    /** Constant for the service parameter. */
+    public static final String SERVICE = "service";
+
+    /** Constant for the theme parameter. */
+    public static final String THEME = "theme";
+
+    /** Constant for the locale parameter. */
+    public static final String LOCALE = "locale";
+
+    /** Constant for the method parameter. */
+    public static final String METHOD = "method";
+
+    /** Supported protocols. */
+    private static final Set<Mechanism> SUPPORTED_PROTOCOLS = ImmutableSet.of(
+            Mechanism.CAS_PROTOCOL,
+            Mechanism.OAUTH_PROTOCOL,
+            Mechanism.OPENID_PROTOCOL,
+            Mechanism.SAML_PROTOCOL,
+            Mechanism.OPENID_CONNECT_PROTOCOL
+    );
+
+    /** The logger. */
+    private final Logger logger = LoggerFactory.getLogger(ClientAction.class);
+
+    /** The clients used for authentication. */
+    @NotNull
+    private final Clients clients;
+
+    /** The service for CAS authentication. */
+    @NotNull
+    private final CentralAuthenticationService centralAuthenticationService;
+
+    /** The handler for osf institution. */
+    @NotNull
+    private final OpenScienceFrameworkInstitutionHandler institutionHandler;
```

Comment:
Need an instance of `OpenScienceFrameworkInstitutionHandler` to access ` OpenScienceFrameworkInstitution` model and dao.